Add a feature flag to allow or disallow writing RTF to the pasteboard when copying/dragging

whsieh opened the pull request:
#### ccdf27571963f7cc3fd1e2559e94c4f74b0494bc
<pre>
Add a feature flag to allow or disallow writing RTF to the pasteboard when copying/dragging
<a href="https://bugs.webkit.org/show_bug.cgi?id=265323">https://bugs.webkit.org/show_bug.cgi?id=265323</a>
<a href="https://rdar.apple.com/118419812">rdar://118419812</a>

Reviewed by Richard Robinson.

On some Cocoa OS versions, the pasteboard will (if necessary) automatically convert HTML and web
archive data written to the pasteboard to RTF, RTFD, or archived `NSAttributedString` data when
pasting in an app that only consumes these native rich text formats.

When this behavior is available (guarded by `HAVE(PASTEBOARD_CONVERSION_FROM_HTML_TO_RTF)`), WebKit
no longer needs to eagerly write RTF, RTFD or attributed string representations of the selected text
to the pasteboard when copying or dragging text, since we can defer to system support for performing
this conversion on paste/drop.

Make some minor adjustments to pasteboard writing logic, to make the attributed string writing part
conditional on this new setting.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::populateRichTextDataIfNeeded):
(WebCore::Editor::writeSelectionToPasteboard):
(WebCore::Editor::writeSelection):

Canonical link: <a href="https://commits.webkit.org/271106@main">https://commits.webkit.org/271106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/984fef77d83e6da010b71b2d8b69a27ec59583d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27352 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29578 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25033 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27820 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24841 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4789 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23478 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4167 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4309 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24474 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30214 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23800 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24894 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30456 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26532 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4337 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2479 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28417 "Found 7 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/document/load-events, /WebKitGTK/TestCookieManager:/webkit/WebKitCookieManager/long-expires, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/selection/listbox, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/selection/menulist, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/value/basic (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5801 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33988 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6587 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4794 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7355 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4718 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/editing/cocoa/EditorCocoa.mm:
```diff
@@ -178,15 +179,17 @@
-    webContent.dataInRTFDFormat = [string containsAttachments] ? dataInRTFDFormat(string.get()) : nullptr;
-    webContent.dataInRTFFormat = dataInRTFFormat(string.get());
-    webContent.dataInAttributedStringFormat = archivedDataForAttributedString(string.get());
+    if (document->settings().writeRichTextDataWhenCopyingOrDragging()) {
```
this could probably be refactored into a generic function since it’s the same few lines as above, but that’s probably overkill.

whsieh replied:
It's not _too_ unwieldy, though I'd need to pull the static `dataInRTFDFormat` / `dataInRTFFormat` helpers out into the `public` section so that I can use them here:

```
template<typename PasteboardContent>
void populateRichTextDataIfNeeded(PasteboardContent& content, const Document& document)
{
    if (!document.settings().writeRichTextDataWhenCopyingOrDragging())
        return;

    auto string = selectionAsAttributedString(document);
    content.dataInRTFDFormat = [string containsAttachments] ? Editor::dataInRTFDFormat(string.get()) : nullptr;
    content.dataInRTFFormat = Editor::dataInRTFFormat(string.get());
    content.dataInAttributedStringFormat = archivedDataForAttributedString(string.get());
}
```

Thanks for the review!

whsieh committed:
Add a feature flag to allow or disallow writing RTF to the pasteboard when copying/dragging
https://bugs.webkit.org/show_bug.cgi?id=265323
rdar://118419812

Reviewed by Richard Robinson.

On some Cocoa OS versions, the pasteboard will (if necessary) automatically convert HTML and web
archive data written to the pasteboard to RTF, RTFD, or archived `NSAttributedString` data when
pasting in an app that only consumes these native rich text formats.

When this behavior is available (guarded by `HAVE(PASTEBOARD_CONVERSION_FROM_HTML_TO_RTF)`), WebKit
no longer needs to eagerly write RTF, RTFD or attributed string representations of the selected text
to the pasteboard when copying or dragging text, since we can defer to system support for performing
this conversion on paste/drop.

Make some minor adjustments to pasteboard writing logic, to make the attributed string writing part
conditional on this new setting.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::populateRichTextDataIfNeeded):
(WebCore::Editor::writeSelectionToPasteboard):
(WebCore::Editor::writeSelection):

Canonical link: https://commits.webkit.org/271106@main
